fix(encoder): 値が飛ぶバグを修正

wakky-alcedo opened the pull request:
値が counter period 分、飛ぶバグを修正しました。
（原因は1周期の間に複数回TIM_FLAGが立っていたときに、それを感知できなかったことだと思われます。）

wakky-alcedo committed:
fix(encoder): 値が飛ぶバグを修正
値が counter period 分、飛ぶバグを修正

wakky-alcedo committed:
refactor(encoder): float 計算をなくし、処理を軽量化

sya-ri commented on tim/encoder.cpp:
```diff
@@ -39,10 +40,10 @@ void Encoder::update() noexcept {
-        if (__HAL_TIM_IS_TIM_COUNTING_DOWN(htim)) {
-            count -= __HAL_TIM_GET_AUTORELOAD(htim);
-        } else {
+        if ((int32_t)(rawCount - lastRawCount) < (int32_t) - __HAL_TIM_GET_AUTORELOAD(htim) / 2) {
```
```suggestion
        if (((int32_t) (lastRawCount - rawCount)) > ((int32_t) (__HAL_TIM_GET_AUTORELOAD(htim) / 2))) { // overflow
```

sya-ri commented on tim/encoder.cpp:
```diff
@@ -39,10 +40,10 @@ void Encoder::update() noexcept {
-        } else {
+        if ((int32_t)(rawCount - lastRawCount) < (int32_t) - __HAL_TIM_GET_AUTORELOAD(htim) / 2) {
             count += __HAL_TIM_GET_AUTORELOAD(htim);
+        } else if ((int32_t)(rawCount - lastRawCount) > (int32_t) + __HAL_TIM_GET_AUTORELOAD(htim) / 2) {
```
```suggestion
        } else if (((int32_t) (rawCount - lastRawCount)) > ((int32_t) (__HAL_TIM_GET_AUTORELOAD(htim) / 2))) { // underflow
```

sya-ri reviewed (changes requested):
条件式をちょっと変えました。等価だと思いますが、試してもらえると嬉しいです

wakky-alcedo committed:
refactor(encoder): overflow, underflow の条件式を修正

wakky-alcedo committed:
fix(encoder): 前回解決しきれていなかった値が飛ぶバグを修正
値が counter period 分、飛ぶバグを修正
原因： 想定していないタイミングで FLAG が立ち、 CLEAR されていたこと
解決策： FLAG を使って overflow / underflow を検知するのを辞めた

sya-ri commented on tim/encoder.cpp:
```diff
@@ -1,6 +1,7 @@
 #ifndef CONFIG_DISABLE_MODULE_TIM
 
 #include "tim/encoder.hpp"
+#include <cmath>
```
```suggestion
```

cmath 使ってなさそう

sya-ri reviewed (changes requested):
確認遅れて申し訳ないです。インクルード消して問題なさそうなら消してほしいです。

wakky-alcedo committed:
refactor: cmath の include を削除

wakky-alcedo commented:
cmath を削除し、build が通ること、書き込んで実行できることを確かめました。
環境がすぐ用意できないため、エンコーダが使えるかどうかは確かめていません。
必要であれば確認しますが、いかがいたしますか？

sya-ri reviewed (approved):
大丈夫です、問題ないと思います
